# Data Preparation

### EDA

In [1]:
import pandas as pd
import numpy as np

In [2]:
os = 'G:\Kumpulan Dataset\Astronomi'

star = pd.read_csv(os+'/map_object_list.csv')

In [3]:
star

type           name        glon       glat  \
0              maser  G351.44+00.65  351.440000   0.650000   
1              maser  G011.49-01.48   11.490000  -1.480000   
2              maser  G014.33-00.64   14.330000  -0.640000   
3              maser  G014.63-00.57   14.630000  -0.570000   
4              maser  G015.03-00.67   15.030000  -0.670000   
...              ...            ...         ...        ...   
2552  ob association        SCO OB1  343.464464   1.389996   
2553  ob association      HD 156154  351.196497   1.369037   
2554  ob association        SCO OB4  352.465040   3.119684   
2555  ob association          R 105  332.997923   1.851210   
2556  ob association        SCO OB2  352.812175  20.187286   

      distance (parsecs)  x (parsecs)  y (parsecs)  z (parsecs)  \
0            1329.000000  1314.111083  -197.802309    15.076703   
1            1250.000000  1224.540716   248.913063   -32.285001   
2            1119.000000  1084.116100   276.942332   -12.499090   
3            1831.000000  1771.545918   462.443798   -18.215177   
4            2004.000000  1935.311186   519.651302   -23.433653   
...                  ...          ...          ...          ...   
2552         1529.261420  1466.016362  -435.243053    37.107179   
2553         1489.704232  1472.154067  -227.993641    35.602076   
2554         1128.317216  1118.574236  -147.957489    61.496259   
2555         2433.451109  2168.180764 -1104.842284    78.651451   
2556          118.045824   117.118139   -14.770179    43.402669   

      radius (parsecs) log10 age (years) Arm (masers)               Source  
0                  NaN               NaN          Sgr  2019ApJ...874...94W  
1                  NaN               NaN          Sgr  2019ApJ...874...94W  
2                  NaN               NaN          Sgr  2019ApJ...874...94W  
3                  NaN               NaN          Sgr  2019ApJ...874...94W  
4                  NaN               NaN          Sgr  2019ApJ...874...94W  
...                ...               ...          ...                  ...  
2552               NaN               NaN          NaN  1978ApJS...38..309H  
2553               NaN               NaN          NaN  1978ApJS...38..309H  
2554               NaN               NaN          NaN  1978ApJS...38..309H  
2555               NaN               NaN          NaN  1978ApJS...38..309H  
2556               NaN               NaN          NaN  1978ApJS...38..309H  

[2557 rows x 12 columns]

In [4]:
star = star[['type','name', 'distance (parsecs)']]

In [5]:
star.isna().sum()

type                  0
name                  0
distance (parsecs)    0
dtype: int64

Data tidak memiliki data NaN dari 2557 data yang tersedia.

In [6]:
df_new = star.copy()

Setelah mengubah data dari list (dengan setiap panjang data ialah 2557) akan dimasukkan kedalam dataframe dalam bentuk dictionary dengan variable "astronom"

### Prepration Dataframe for modeling

In [7]:
typee = df_new['type'].to_list()
name = df_new['name'].to_list()
distance = df_new['distance (parsecs)'].to_list()

In [8]:
print(len(typee))
print(len(name))
print(len(distance))

2557
2557
2557


In [9]:
astronom = pd.DataFrame({
    'type' : typee,
    'name' : name,
    'distance' : distance
})

In [10]:
astronom

type           name     distance
0              maser  G351.44+00.65  1329.000000
1              maser  G011.49-01.48  1250.000000
2              maser  G014.33-00.64  1119.000000
3              maser  G014.63-00.57  1831.000000
4              maser  G015.03-00.67  2004.000000
...              ...            ...          ...
2552  ob association        SCO OB1  1529.261420
2553  ob association      HD 156154  1489.704232
2554  ob association        SCO OB4  1128.317216
2555  ob association          R 105  2433.451109
2556  ob association        SCO OB2   118.045824

[2557 rows x 3 columns]

### TF-IDF dan Cosine Similarity

Disini akan melakukan **TF-IDF** untuk menghitung bobot dari setiap tipe menggunakan library sklearn

In [11]:
from sklearn.feature_extraction.text import TfidfVectorizer

Membuat variable baru dengan nama data yang menyimpan dataframe sebelumnya dan variable **tf** sebagai fungsi TF-IDF

In [12]:
data = astronom

In [13]:
tf = TfidfVectorizer()

Melakukan extract dan get feature dari tipe planet yang ada pada data

In [14]:
tf.fit(data['type'])

tf.get_feature_names()

C:\Users\adiba\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


['association', 'cluster', 'hii', 'maser', 'ob', 'open', 'region']

Setelah melakukan get feature yang tersedia, maka ditransformasikan menjadi kolom dan apabila di shape akan mendapatkan menjadi 2557 baris dan 7 kolom

In [15]:
tfidf_matrix = tf.fit_transform(data['type'])

tfidf_matrix.shape

(2557, 7)

In [16]:
tfidf_matrix.todense()

matrix([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ],
        [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ],
        [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ],
        ...,
        [0.70710678, 0.        , 0.        , ..., 0.70710678, 0.        ,
         0.        ],
        [0.70710678, 0.        , 0.        , ..., 0.70710678, 0.        ,
         0.        ],
        [0.70710678, 0.        , 0.        , ..., 0.70710678, 0.        ,
         0.        ]])

In [17]:
pd.DataFrame(
            tfidf_matrix.todense(),
            columns=tf.get_feature_names(),
            index = data.name
            )

association  cluster  hii  maser        ob  open  region
name                                                                   
G351.44+00.65     0.000000      0.0  0.0    1.0  0.000000   0.0     0.0
G011.49-01.48     0.000000      0.0  0.0    1.0  0.000000   0.0     0.0
G014.33-00.64     0.000000      0.0  0.0    1.0  0.000000   0.0     0.0
G014.63-00.57     0.000000      0.0  0.0    1.0  0.000000   0.0     0.0
G015.03-00.67     0.000000      0.0  0.0    1.0  0.000000   0.0     0.0
...                    ...      ...  ...    ...       ...   ...     ...
SCO OB1           0.707107      0.0  0.0    0.0  0.707107   0.0     0.0
HD 156154         0.707107      0.0  0.0    0.0  0.707107   0.0     0.0
SCO OB4           0.707107      0.0  0.0    0.0  0.707107   0.0     0.0
R 105             0.707107      0.0  0.0    0.0  0.707107   0.0     0.0
SCO OB2           0.707107      0.0  0.0    0.0  0.707107   0.0     0.0

[2557 rows x 7 columns]

In [18]:
from sklearn.metrics.pairwise import cosine_similarity

In [19]:
cosine_sim = cosine_similarity(tf.fit_transform(data['type']))
cosine_sim

array([[1., 1., 1., ..., 0., 0., 0.],
       [1., 1., 1., ..., 0., 0., 0.],
       [1., 1., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 1., 1.],
       [0., 0., 0., ..., 1., 1., 1.],
       [0., 0., 0., ..., 1., 1., 1.]])

In [20]:
cosine_sim_df = pd.DataFrame(cosine_sim, columns = data['name'], index = data['name'])
print('Shape:', cosine_sim_df.shape)

Shape: (2557, 2557)


In [21]:
cosine_sim_df.sample(10, axis=1).sample(5, axis=0)

name                                  King_15  Collinder_261  UBC 106  \
name                                                                    
NGC_129                                   1.0            1.0      1.0   
UBC 539                                   1.0            1.0      1.0   
UBC 341                                   1.0            1.0      1.0   
Sh 2-30 (Trifid nebula | Messier 20)      0.0            0.0      0.0   
UBC 642                                   1.0            1.0      1.0   

name                                  Czernik_30  Gulliver_7  Ferrero_1  \
name                                                                      
NGC_129                                      1.0         1.0        1.0   
UBC 539                                      1.0         1.0        1.0   
UBC 341                                      1.0         1.0        1.0   
Sh 2-30 (Trifid nebula | Messier 20)         0.0         0.0        0.0   
UBC 642                                      1.0         1.0        1.0   

name                                  UPK_45  Liu_Pang_2198  UBC 655  NGC_2428  
name                                                                            
NGC_129                                  1.0            1.0      1.0       1.0  
UBC 539                                  1.0            1.0      1.0       1.0  
UBC 341                                  1.0            1.0      1.0       1.0  
Sh 2-30 (Trifid nebula | Messier 20)     0.0            0.0      0.0       0.0  
UBC 642                                  1.0            1.0      1.0       1.0

# Modeling 

In [22]:
def planet_recommendations(name, similarity_data=cosine_sim_df, items= data[['name', 'type', 'distance']], k=10):
    index = similarity_data.loc[:,name].to_numpy().argpartition(
        range(-1, -k, -1))
    
    closest = similarity_data.columns[index[-1:-(k+2):-1]]
    
    closest = closest.drop(name, errors='ignore')
    
    return pd.DataFrame(closest).merge(items).head(k)

In [23]:
data[data.name.eq('UBC 119')]

type     name     distance
1735  open cluster  UBC 119  1781.261133

# Evaluation

In [24]:
result = planet_recommendations('UBC 119')
result

name          type  distance
0  Skiff_J0614+12.9  open cluster    3601.6
1          NGC_6791  open cluster    4530.8
2          NGC_6756  open cluster    3220.6
3          NGC_6755  open cluster    2471.2
4          NGC_6735  open cluster    2247.9
5          NGC_6728  open cluster    1770.8
6          NGC_6793  open cluster     588.8
7          NGC_6709  open cluster    1064.4
8          NGC_6633  open cluster     391.6
9          NGC_6704  open cluster    2158.2

In [25]:
relevant = len(result[result.type == 'open cluster'])
not_relevant = len(result[~(result.type == 'open cluster')])
items_total = len(result.type)

In [26]:
accuracy = ((relevant + not_relevant)/items_total) * 100
precision = (relevant / items_total) * 100
recall = (relevant/(relevant + not_relevant)) * 100

print('Accuracy: {}%'.format(round(accuracy),0))
print('Precision: {}%'.format(round(precision),0))
print('Recall: {}%'.format(round(recall),0))

Accuracy: 100%
Precision: 100%
Recall: 100%
